### ISSUE TITLE
IWriteBatch.Update() does not support DateTimeOffset

### ISSUE THREAD

**Issue by barrysohl** [TobiasBuchholz/Plugin.Firebase#340]
When calling Plugin.Firebase.Firestore `IWriteBatch.UpdateData( IDocumentReference, Dictionary<object,object>)`, having a DateTimeOffset value in the Dictionary does not seem to be supported. The following exception is thrown:

`System.ArgumentException: Do not know how to marshal object of type 'System.DateTimeOffset' to an NSObject`

Am I doing something wrong, or does this functionality need to be added? Thanks in advance!

**Comment by barrysohl** [comment]
Pinging this ticket with more info:

1. DateTimeOffset works fine with `UpdateDataAsync()`
2. DateTimeOffset works fine with `IWriteBatch.SetData()`
3. DateTimeOffset still does NOT work with `IWriteBatch.UpdateData()`

Can anyone provide more information please? I am having to unnecessarily use SetData as a workaround where UpdateData would be much more efficient.

**Comment by AdamEssenmacher** [comment]
A PR from the community would be welcome on this one.

**Comment by barrysohl** [comment]
We will gladly donate $$ to whoever can fix this one! It is a showstopper for us currently. Unfortunately, I don't think we have anyone in-house comfortable with requesting the pull themselves.

**Comment by AdamEssenmacher** [comment]
> We will gladly donate $$ to whoever can fix this one! It is a showstopper for us currently. Unfortunately, I don't think we have anyone in-house comfortable with requesting the pull themselves.

I do have one-time sponsorship options on my GH profile. I'm sure I could find a fix within an afternoon.

**Comment by barrysohl** [comment]
Done! I have a feeling it may be as simple as the switch statement that maps Dictionary types to a value converter is missing DateTimeOffset. If it ends-up being a quick fix, there's another ticket we would love some help on :)

**Comment by barrysohl** [comment]
#359 

There are other tickets as well, but we have found workarounds for most everything else. Thanks!

**Comment by AdamEssenmacher** [comment]
For some context, a lot of what the FIrestore plugin does is translate between .NET objects and their native counterparts. The native Firestore SDKs only understand ObjectiveC or Java. So, for example, we cannot pass a .NET DateTimeOffset object directly to the native SDK--we need to first convert it to an `NSDate` or Java `Date` first. The purpose of the Firestore plugin is to abstract this native translation from the developer, so you can use normal .NET types.

The reported bug is only occurring on iOS. Also, the overload `UpdateData(IDocumentReference document, params (string, object)[] data)` works correctly. This means you can work around this problem by doing something like:

```
batch.UpdateData(doc, dictionary.Select(x => (x.Key.ToString(), x.Value)).ToArray());
```
This approach could be worked into an extension method like:
```
public static class FirestoreExtensions
{
    public static IWriteBatch UpdateDataHack(this IWriteBatch batch, IDocumentReference document, Dictionary<object, object> data)
    {
        return batch.UpdateData(document, data.Select(x => (x.Key.ToString(), x.Value)).ToArray());
    }
}
```

As for an actual _fix_, #381 should do it whenever Tobi publishes the next release.

**Comment by barrysohl** [comment]
Thanks for getting this fixed!!

I figured it was just a missing DateTimeOffset conversion somewhere. I assume this is also related to why some complex types (such as an IDictionary with ILIst objects in it) do not work, where they worked for us previously in f-miyu's Plugin.CloudFirestore? We were able to work around most of these issues by flattening our data model, but that resulted in losing some backward compatibility.
